### PR TITLE
[TT-9327] Decoding the URL request first, before handling any additional logic

### PR DESCRIPTION
**Pull request by lghiur**
this path works: /payment-intents
but this path doesn't: /payment%2Dintents

Encoded URLs aren't being rewritten when URL rewrite is applied.

One edge case scenario that could break backwards compatibility (as described by @buger ), is that users can rely on escaped characters, and try to match them from the the url rewrite rules.

In order to accomodate that, we are running url rewrite middleware twice:
- once on the raw path
-  if transformations are failing and the url contains encoded characters, then we run it second time, with decoded URL

## How This Has Been Tested

Unit test and manually

## Types of changes

- [√ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] I ensured that the documentation is up to date
- [ ] I explained why this PR updates go.mod in detail with reasoning why it's required
- [ ] I would like a code coverage CI quality gate exception and have explained why

### CODE REVIEW

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5345/merge
Commit:  
Triggered by: pull_request (@lghiur)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5645815209)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5345/merge
Commit: af476766a1c5e1c5b456662c7a80ec0bbc68b404 
Triggered by: pull_request (@lghiur)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5645850234)

File: gateway/mw_url_rewrite.go [hunk 1]
```diff
@@ -39,7 +39,13 @@ var metaMatch = regexp.MustCompile(`\$tyk_meta.([A-Za-z0-9_\-\.]+)`)
 var secretsConfMatch = regexp.MustCompile(`\$secret_conf.([A-Za-z0-9[.\-\_]+)`)
 
 func (gw *Gateway) urlRewrite(meta *apidef.URLRewriteMeta, r *http.Request) (string, error) {
-	path := r.URL.String()
+	rawPath := r.URL.String()
+
+	path, err := url.PathUnescape(rawPath)
+	if err != nil {
+		return rawPath, fmt.Errorf("Error decoding URL path: %s", rawPath)
```
Comment (furkansenharputlu) [inline gateway/mw_url_rewrite.go:46]: failed to decode..

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5345/merge
Commit: ba6e0b9e8ec5341bdf6a7c8702145c3c10018cba 
Triggered by: pull_request (@lghiur)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5646405122)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5345/merge
Commit: 24a82b0843817facd5c974b08f835e0e6cc2a22e 
Triggered by: pull_request (@lghiur)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5646415132)

**Comment by buger** [comment]
Backward compatiblity issue can happen when users depended on URL encoding in the url rewrite. And now it will break. 
Proposal: check if url rewrite contains escaped characters, if yes, do not un-escape, otherwise do it.

**Comment by lghiur** [comment]
> Backward compatiblity issue can happen when users depended on URL encoding in the url rewrite. And now it will break. Proposal: check if url rewrite contains escaped characters, if yes, do not un-escape, otherwise do it.

Solved in a follow up commit, where we could potentially run the url rewrite middleware twice. (suggested by @buger )

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5345/merge
Commit: 05c5c8d6075e067d4236cbfa5f82f92467c082c4 
Triggered by: pull_request (@lghiur)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5648946611)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5345/merge
Commit: e78a5ac87f7aede08ae45d775d78ec7bc512b313 
Triggered by: pull_request (@lghiur)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5648958242)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5345/merge
Commit: 5b9dde418166429bd499bdc698ace2577f22ee07 
Triggered by: pull_request (@lghiur)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5649158454)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5345/merge
Commit: 284853a0012844b048b87cce3ebb47cb47ed46dd 
Triggered by: pull_request (@lghiur)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5653109864)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5345/merge
Commit: 0742f5ef3080bf9ba8d0b527a19d41f34fadb8a5 
Triggered by: pull_request (@lghiur)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5653210058)

**Review by jeffy-mathew — APPROVED** [review]

File: gateway/mw_url_rewrite.go [hunk 2]
```diff
@@ -38,8 +38,18 @@ var envValueMatch = regexp.MustCompile(`\$secret_env.([A-Za-z0-9_\-\.]+)`)
 var metaMatch = regexp.MustCompile(`\$tyk_meta.([A-Za-z0-9_\-\.]+)`)
 var secretsConfMatch = regexp.MustCompile(`\$secret_conf.([A-Za-z0-9[.\-\_]+)`)
 
-func (gw *Gateway) urlRewrite(meta *apidef.URLRewriteMeta, r *http.Request) (string, error) {
-	path := r.URL.String()
+func (gw *Gateway) urlRewrite(meta *apidef.URLRewriteMeta, r *http.Request, decodeUrl bool) (string, error) {
```
Comment (jeffy-mathew) [inline gateway/mw_url_rewrite.go:41]: can we use `decodeURL` here?

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5345/merge
Commit: 06be0e227163e87b1b1891ec67849632bc3fa9d9 
Triggered by: pull_request (@lghiur)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5655125207)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5345/merge
Commit: c7e82467581fcb032b5ee18c4c2ee2eb0d1828ad 
Triggered by: pull_request (@lghiur)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5655452967)

**Pull request merged**